File: BLE Test/UARTPeripheral.m
//
//  UARTPeripheral.m
//  nRF UART
//
//  Created by Ole Morten on 1/12/13.
//  Copyright (c) 2013 Nordic Semiconductor. All rights reserved.
//

#import <CoreBluetooth/CoreBluetooth.h>
#import "UARTPeripheral.h"
#import "CBUUID+StringExtraction.h"

@interface UARTPeripheral ()
@property CBService *uartService;
@property CBCharacteristic *rxCharacteristic;
@property CBCharacteristic *txCharacteristic;

@end

@implementation UARTPeripheral
@synthesize peripheral = _peripheral;
@synthesize delegate = _delegate;

@synthesize uartService = _uartService;
@synthesize rxCharacteristic = _rxCharacteristic;
@synthesize txCharacteristic = _txCharacteristic;


#pragma mark - UUID Retrieval


+ (CBUUID*)uartServiceUUID{
    
    return [CBUUID UUIDWithString:@"6e400001-b5a3-f393-e0a9-e50e24dcca9e"];
}


+ (CBUUID*)txCharacteristicUUID{
    
    return [CBUUID UUIDWithString:@"6e400002-b5a3-f393-e0a9-e50e24dcca9e"];
}


+ (CBUUID*)rxCharacteristicUUID{
    
    return [CBUUID UUIDWithString:@"6e400003-b5a3-f393-e0a9-e50e24dcca9e"];
}


+ (CBUUID*)deviceInformationServiceUUID{
    
    return [CBUUID UUIDWithString:@"180A"];
}


+ (CBUUID*)hardwareRevisionStringUUID{
    
    return [CBUUID UUIDWithString:@"2A27"];
}


#pragma mark - Utility methods


- (UARTPeripheral*)initWithPeripheral:(CBPeripheral*)peripheral delegate:(id<UARTPeripheralDelegate>) delegate{
    
    if (self = [super init]){
        self.peripheral = peripheral;
        self.peripheral.delegate = self;
        self.delegate = delegate;
    }
    return self;
}


- (void)didConnect{
    
    if(_peripheral.services){
        printf("Skipping service discovery for %s\r\n", [_peripheral.name UTF8String]);
        [self peripheral:_peripheral didDiscoverServices:nil]; //already discovered services, DO NOT re-discover. Just pass along the peripheral.
        return;
    }
    
    printf("Starting service discovery for %s\r\n", [_peripheral.name UTF8String]);
    
    [_peripheral discoverServices:@[self.class.uartServiceUUID, self.class.deviceInformationServiceUUID]];
    
}


- (void)didDisconnect{
    
}


- (void)writeString:(NSString*)string{
    
    NSData *data = [NSData dataWithBytes:string.UTF8String length:string.length];
    
    [self writeRawData:data];
}


- (void)writeRawData:(NSData*)data{
    
//    NSLog(@"writeRawData:");
    
    if ((self.txCharacteristic.properties & CBCharacteristicPropertyWriteWithoutResponse) != 0){
        
        [self.peripheral writeValue:data forCharacteristic:self.txCharacteristic type:CBCharacteristicWriteWithoutResponse];
    }
    else if ((self.txCharacteristic.properties & CBCharacteristicPropertyWrite) != 0){
        [self.peripheral writeValue:data forCharacteristic:self.txCharacteristic type:CBCharacteristicWriteWithResponse];
    }
    else{
        NSLog(@"No write property on TX characteristic, %d.", self.txCharacteristic.properties);
    }
    
}


- (BOOL)compareID:(CBUUID*)firstID toID:(CBUUID*)secondID{
    
    if ([[firstID representativeString] compare:[secondID representativeString]] == NSOrderedSame) {
        return YES;
    }
    
    else
        return NO;
    
}


- (void)setupPeripheralForUse:(CBPeripheral*)peripheral{
    
    printf("Set up peripheral for use");
    
    for (CBService *s in peripheral.services) {
        
        for (CBCharacteristic *c in [s characteristics]){
            
            if ([self compareID:c.UUID toID:self.class.rxCharacteristicUUID]){
                
                printf("Found RX characteristic\r\n");
                self.rxCharacteristic = c;
                
                [self.peripheral setNotifyValue:YES forCharacteristic:self.rxCharacteristic];
            }
            
            else if ([self compareID:c.UUID toID:self.class.txCharacteristicUUID]){
                
                printf("Found TX characteristic\r\n");
                self.txCharacteristic = c;
            }
            
            else if ([self compareID:c.UUID toID:self.class.hardwareRevisionStringUUID]){
                
                printf("Found Hardware Revision String characteristic\r\n");
                [self.peripheral readValueForCharacteristic:c];
                
                //HW characteristic is last to be discovered, notify delegate we're all set
                [_delegate uartDidConnect];
                
            }
            
        }

    }
    
}


#pragma mark - CBPeripheral Delegate methods


- (void)peripheral:(CBPeripheral*)peripheral didDiscoverServices:(NSError*)error{
    
    printf("Did Discover Services\r\n");
    
    if (!error) {
        
        for (CBService *s in [peripheral services]){
            
            if (s.characteristics){
                [self peripheral:peripheral didDiscoverCharacteristicsForService:s error:nil]; //already discovered characteristic before, DO NOT do it again
            }
            
            else if ([self compareID:s.UUID toID:self.class.uartServiceUUID]){
                
                printf("Found correct service\r\n");
                
                self.uartService = s;
                
                [self.peripheral discoverCharacteristics:@[self.class.txCharacteristicUUID, self.class.rxCharacteristicUUID] forService:self.uartService];
            }
            
            else if ([self compareID:s.UUID toID:self.class.deviceInformationServiceUUID]){
                
                [self.peripheral discoverCharacteristics:@[self.class.hardwareRevisionStringUUID] forService:s];
            }
        }
    }
    
    else{
        
        printf("Error discovering services\r\n");
        
        [_delegate uartDidEncounterError:@"Error discovering services"];
        
        return;
    }
    
}


- (void)peripheral:(CBPeripheral*)peripheral didDiscoverCharacteristicsForService:(CBService*)service error:(NSError*)error{
    
//    NSLog(@"Did Discover Characteristics");
    
    if (!error){
        
        CBService *s = [peripheral.services objectAtIndex:(peripheral.services.count - 1)];
        if([self compareID:service.UUID toID:s.UUID]){
            
            //last service discovered
            printf("Found all characteristics\r\n");
            
            [self setupPeripheralForUse:peripheral];
            
        }
        
    }
    
    else{
        
        printf("Error discovering characteristics: %s\r\n", [error.description UTF8String]);
        
        [_delegate uartDidEncounterError:@"Error discovering characteristics"];
        
        return;
    }
    
}


- (void)peripheral:(CBPeripheral*)peripheral didUpdateValueForCharacteristic:(CBCharacteristic*)characteristic error:(NSError*)error{
    
//    NSLog(@"Received data on a characteristic.");
    

    if (!error){
        if (characteristic == self.rxCharacteristic){
            
            [self.delegate didReceiveData:[characteristic value]];
        }
        
        else if ([self compareID:characteristic.UUID toID:self.class.hardwareRevisionStringUUID]){
            
            NSString *hwRevision = @"";
            const uint8_t *bytes = characteristic.value.bytes;
            for (int i = 0; i < characteristic.value.length; i++){
                
                hwRevision = [hwRevision stringByAppendingFormat:@"0x%x, ", bytes[i]];
            }
            
            [self.delegate didReadHardwareRevisionString:[hwRevision substringToIndex:hwRevision.length-2]];
        }
    }
    
    else{
        
        printf("Error receiving notification for characteristic %s: %s\r\n", [characteristic.description UTF8String], [error.description UTF8String]);
        
        [_delegate uartDidEncounterError:@"Error receiving notification for characteristic"];
        
        return;
    }
    
}


@end

File: BLE Test/UARTPeripheral.swift
//
//  UARTPeripheral.swift
//  Adafruit Bluefruit LE Connect
//
//  Created by Collin Cunningham on 10/12/14.
//  Copyright (c) 2014 Adafruit Industries. All rights reserved.
//

import Foundation
import CoreBluetooth

@objc protocol UARTPeripheralDelegate: Any {
    
    func didReceiveData(newData:NSData)
    func didReadSoftwareRevisionString(string:NSString)
    func uartDidEncounterError(error:NSString)
    
}

class UARTPeripheral: NSObject, CBPeripheralDelegate {
    
    var currentPeripheral:CBPeripheral!
    var delegate:UARTPeripheralDelegate!
    var uartService:CBService?
    var rxCharacteristic:CBCharacteristic?
    var txCharacteristic:CBCharacteristic?
    
    
    //MARK: UUID Retrieval
    
    class func uartServiceUUID()->CBUUID{
    
        return CBUUID.UUIDWithString("6e400001-b5a3-f393-e0a9-e50e24dcca9e")
        
    }
    
    
    class func txCharacteristicUUID()->CBUUID{
    
        return CBUUID.UUIDWithString("6e400002-b5a3-f393-e0a9-e50e24dcca9e")
    }
    
    
    class func rxCharacteristicUUID()->CBUUID{
    
        return CBUUID.UUIDWithString("6e400003-b5a3-f393-e0a9-e50e24dcca9e")
    }
    
    
    class func deviceInformationServiceUUID()->CBUUID{
    
        return CBUUID.UUIDWithString("180A")
    }
    
    
    class func hardwareRevisionStringUUID()->CBUUID{
    
        return CBUUID.UUIDWithString("2A27")
    }
    
    
    class func manufacturerNameStringUUID()->CBUUID{    //    Manufacturer Name String	0x2A29
        
        return CBUUID.UUIDWithString("2A29")
    }
    
    
    class func modelNumberStringUUID()->CBUUID{         //    Model Number String		0x2A24
        
        return CBUUID.UUIDWithString("2A24")
    }
    
    
    class func firmwareRevisionStringUUID()->CBUUID{    //    Firmware Revision String	0x2A26
        
        return CBUUID.UUIDWithString("2A26")
    }
    
    
    class func softwareRevisionStringUUID()->CBUUID{    //    Software Revision String  0x2A28
        
        return CBUUID.UUIDWithString("2A28")
    }
    
    
    //MARK: Utility methods
    
    init(peripheral:CBPeripheral, delegate:UARTPeripheralDelegate){
        
        super.init()
        
        self.currentPeripheral = peripheral
        self.currentPeripheral.delegate = self
        self.delegate = delegate
    
    }
    
    
    func didConnect() {
    
    //Respond to peripheral connection
    
        if currentPeripheral.services != nil{
            println("Skipping service discovery for \(currentPeripheral.name.utf8)")
            peripheral(currentPeripheral, didDiscoverServices: nil)  //already discovered services, DO NOT re-discover. Just pass along the peripheral.
            return
        }
    
        println("Starting service discovery for \(currentPeripheral.name.utf8)")

        currentPeripheral.discoverServices([UARTPeripheral.uartServiceUUID(), UARTPeripheral.deviceInformationServiceUUID()])
//        currentPeripheral.discoverServices(nil)
    
    }
    
    
    func didDisconnect() {
    
    //Respond to peripheral disconnection
    
    }
    
    
    func writeString(string:NSString){
    
    //Send string to peripheral
    
        let data = NSData(bytes: string.UTF8String, length: string.length)
        
        writeRawData(data)
    }
    
    
    func writeRawData(data:NSData) {
    
    //Send data to peripheral
        
        if (txCharacteristic == nil){
            println("WARNING: Unable to write data without txcharacteristic!")
            return
        }
    
        if (txCharacteristic!.properties.toRaw() & CBCharacteristicProperties.WriteWithoutResponse.toRaw()) != 0 {
            
            currentPeripheral.writeValue(data, forCharacteristic: self.txCharacteristic, type: CBCharacteristicWriteType.WithoutResponse)
        }
        
        else if ((txCharacteristic!.properties.toRaw() & CBCharacteristicProperties.Write.toRaw()) != 0){
            currentPeripheral.writeValue(data, forCharacteristic: txCharacteristic, type: CBCharacteristicWriteType.WithResponse)
        }
            
        else{
            println("No write property on TX characteristic, \(txCharacteristic!.properties)")
        }
    
    }
    
    
    func UUIDsAreEqual(firstID:CBUUID, _ secondID:CBUUID)->Bool {
        
        if firstID.representativeString() == secondID.representativeString() {
            return true
        }
            
        else {
            return false
        }
        
    }
    
    
    func setupPeripheralForUse(peripheral:CBPeripheral) {
        
        println("Set up peripheral for use")
        
        let services = peripheral.services as [CBService]
        
        for s in services {
            
            let characteristics = s.characteristics as [CBCharacteristic]
            
            for c in characteristics {
                
                if UUIDsAreEqual(c.UUID, UARTPeripheral.rxCharacteristicUUID()) {
                    
                    println("Found RX characteristic")
                    rxCharacteristic = c
                    peripheral.setNotifyValue(true, forCharacteristic: rxCharacteristic)
                    
                }
                    
                else if UUIDsAreEqual(c.UUID, UARTPeripheral.txCharacteristicUUID()) {
                    
                    println("Found TX characteristic")
                    txCharacteristic = c
                }
                    
                else if UUIDsAreEqual(c.UUID, UARTPeripheral.hardwareRevisionStringUUID()) {
                    
                    println("Found Hardware Revision String characteristic")
                    peripheral.readValueForCharacteristic(c)  //Once hardware revision string is read connection will be complete …
                    
                }
                
            }
            
        }
        
    }
    
    
    //MARK: CBPeripheral Delegate methods
    
    func peripheral(peripheral: CBPeripheral!, didDiscoverServices error: NSError!) {
        
        //Respond to finding a new service on peripheral
        
        if error != nil {
            
            handleError("Error discovering services")
            
            return
        }
        
        println("Did Discover Services")
        
        
        let services = peripheral.services as [CBService]
        
        for s in services {
            
            if (s.characteristics != nil){
                
                self.peripheral(peripheral, didDiscoverCharacteristicsForService: s, error: nil)    // If characteristics have already been discovered, do not check again
                
            }
                
            else if UUIDsAreEqual(s.UUID, UARTPeripheral.uartServiceUUID()) {
                
                println("Found Service: UART")
                
                uartService = s
//                peripheral.discoverCharacteristics([UARTPeripheral.txCharacteristicUUID(), UARTPeripheral.rxCharacteristicUUID()], forService: uartService)
                peripheral.discoverCharacteristics(nil, forService: uartService)
                
            }
                
            else if UUIDsAreEqual(s.UUID, UARTPeripheral.deviceInformationServiceUUID()){
                
                println("Found Service: Device Information")
                
                peripheral.discoverCharacteristics(nil, forService: s)
            }
        }
        
    }
    
    
    func peripheral(peripheral: CBPeripheral!, didDiscoverCharacteristicsForService service: CBService!, error: NSError!) {
        
        //Respond to finding a new characteristic on service
        
        if error != nil {
            handleError("Error discovering characteristics")
            
            return
        }
        
        println("Service \(service.description) has \(service.characteristics.count) characteristics")
        
//                if service.UUID == UARTPeripheral.deviceInformationServiceUUID() && service.characteristics.count == 0 {    //MARK: Hack for no HW Revision
//                    println("Device Information Service contains no characteristics, skipping HW Rev")
//                    delegate.didReadHardwareRevisionString("NONE")  //No HW Rev, connect anyway
//                }
        
        for c in (service.characteristics as [CBCharacteristic]) {
            
            switch c.UUID {
            case UARTPeripheral.rxCharacteristicUUID():         //"6e400003-b5a3-f393-e0a9-e50e24dcca9e"
                println("Found Characteristic: RX")
                rxCharacteristic = c
                currentPeripheral.setNotifyValue(true, forCharacteristic: rxCharacteristic)
                break
            case UARTPeripheral.txCharacteristicUUID():         //"6e400002-b5a3-f393-e0a9-e50e24dcca9e"
                println("Found Characteristic: TX")
                txCharacteristic = c
                break
            case UARTPeripheral.hardwareRevisionStringUUID():   //"2A27"
                println("Found Characteristic: Hardware Revision")
                currentPeripheral.readValueForCharacteristic(c)
                break
            case UARTPeripheral.manufacturerNameStringUUID():
                println("Found Characteristic: Manufacturer Name")
                currentPeripheral.readValueForCharacteristic(c)
                break
            case UARTPeripheral.modelNumberStringUUID():
                println("Found Characteristic: Model Number")
                currentPeripheral.readValueForCharacteristic(c)
                break
            case UARTPeripheral.firmwareRevisionStringUUID():
                println("Found Characteristic: Firmware Revision")
                currentPeripheral.readValueForCharacteristic(c)
                break
            case UARTPeripheral.softwareRevisionStringUUID():
                println("Found Characteristic: Software Revision")
                currentPeripheral.readValueForCharacteristic(c)
                break
            default:
                println("Found Characteristic: Unknown")
                break
            }
            
        }
        
    }
    
    
    func peripheral(peripheral: CBPeripheral!, didUpdateValueForCharacteristic characteristic: CBCharacteristic!, error: NSError!) {
        
        //Respond to value change on peripheral
        
        if error != nil {
            handleError("Error receiving notification for characteristic\(characteristic.description.utf8): \(error.description.utf8)")
            return
        }
        
        if (characteristic == self.rxCharacteristic){
            
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                self.delegate.didReceiveData(characteristic.value)
            })
            
        }
            
        else if UUIDsAreEqual(characteristic.UUID, UARTPeripheral.softwareRevisionStringUUID()) {
            
            var swRevision = NSString(string: "")
            let bytes:UnsafePointer<Void> = characteristic.value.bytes
            //    const uint8_t *bytes = characteristic.value.bytes;
            
            for i in 0...characteristic.value.length {  //TODO: Check
                
                swRevision = NSString(format: "0x%x", UInt8(bytes[i]) ) //TODO: Check
            }
            
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                self.delegate.didReadSoftwareRevisionString(swRevision)
            })
        }
        
        
    }
    
    
    func peripheral(peripheral: CBPeripheral!, didDiscoverIncludedServicesForService service: CBService!, error: NSError!) {
        
        //Respond to finding a new characteristic on service
        
        if error != nil {
            handleError("Error discovering included service characteristics")
            return
        }
        
        println("Included service: \(service.description) has \(service.includedServices.count) included services")
        
//        if service.characteristics.count == 0 {
//            currentPeripheral.discoverIncludedServices(nil, forService: service)
//        }
        
        for s in (service.includedServices as [CBService]) {
            
            println("   included service: \(s.description)")
        }
        
    }
    
    
    func handleError(errorString:String) {
        
        println(errorString)
        
        dispatch_async(dispatch_get_main_queue(), { () -> Void in
            self.delegate.uartDidEncounterError(errorString)
        })
        
    }
    
    
}